securesystemslib/metadata.py

```diff
@@ -97,3 +98,50 @@ def pae(self) -> bytes:
             len(self.payload),
             self.payload,
         )
+
+    def sign(self, signer: Signer) -> Signature:
+        """sign the payload and create the signature
+
+        Arguments:
+            signer: A "Signer" class instance
+
+        Raises:
+            TypeError: If "signer" is not a instance of "Signer" class.
+
+        Returns:
+            A "Signature" instance.
+        """
+
+        if not isinstance(signer, Signer):
+            raise TypeError(f"expected type 'Signer', got {type(signer).__name__}")
+
+        signature = signer.sign(self.pae)
+        self.signatures.append(signature)
```

Comment:
Before appending maybe you will want to check if this signature doesn't exist? 
Does a `set` makes sense for `Envelope.signatures`?

Reply:
Not sure about it, according to current signature wrappers:
in-toto: signatures is a list and appends the signatures.
python-tuf: signatures is a dict and appends with `keyid`.

Also, in DSSE `keyid` is optional. But what I can do is:
```suggestion
        if signature.keyid:
            for s in self.signatures:
                if s.keyid == signature.keyid:
                    self.signatures.remove(s)
                    break
        self.signatures.append(signature)
```

Reply:
> in-toto: signatures is a list and appends the signatures.
Doesn't `in-toto` make a check if a signature already exists?

In `python-tuf` to add a new signature you should use `sign`: https://github.com/theupdateframework/python-tuf/blob/8a03abfdeb5a6b6bc892859f65425be0834893a8/tuf/api/metadata.py#L342.
In `sign` there is an `append` option which if used will add an additional signature.
If `append` is True and a key is already used to sign a file, then the signature will exist in the dictionary will be replaced by a new `Signature` instance with the same content: https://github.com/theupdateframework/python-tuf/blob/8a03abfdeb5a6b6bc892859f65425be0834893a8/tuf/api/metadata.py#L389.
This removes the possibility of two existing `Signature` instances with the same content.

The question is do we want to allow duplicating signatures?
I think duplicating signatures could be a problem because they artificially increase the numbers of signatures used and thus `threshold` loses its importance.
I don't think an attacker can utilize that as for that he/she will need the private key used for signing (and if he has that he can break the system on multiple levels), so maybe I shouldn't worry.

@lukpueh any thoughts?

PS: If you want to remove an already existing signature you can directly use `==` and not relly if a `keyid` exists in a signature as a `__eq__` implementation exist: https://github.com/in-toto/securesystemslib/blob/d3f48889595c6252ad30ccbec2b4c7d87795e37d/securesystemslib/signer.py#L40

Reply:
> Doesn't `in-toto` make a check if a signature already exists?

I don't think in-toto's metadata wrapper do that.

@adityasaky whats your opinion about this.

Reply:
in-toto does allow multiple signatures from the same key on a particular document but that's because multiple signatures only really apply to the layout file. For a threshold k out of n keys for a particular step, you're not going to have one link with multiple signatures, but rather k different link files. We verify here to see we have k distinct link metadata files: https://github.com/in-toto/in-toto/blob/f6a4f6239ef28fc482adf147af79f030086f63f0/in_toto/verifylib.py#L506-L507. As for the layout, verification is dependent on each submitted public key.

Reply:
I still think that in order to escape possible problems regarding threshold computation it's better if we don't allow duplicating signatures. They just don't make sense to exist.
@adityasaky  maybe before merging, we can wait for @lukpueh input here?

Reply:
We could, as long as we don't block @PradyumnaKrishna. I maintain that we shouldn't be getting too involved with duplicate signatures in the DSSE implementation. Perhaps this is even a hint that we shouldn't get too involved in the threshold aspects in the envelope either as each consumer may have different takes on how to handle thresholds. Looking for duplicate signatures gets messy here, IMO, for a number of reasons:
1. DSSE doesn't specify how keyids are supposed to be calculated so a key could theoretically be presented with multiple keyids, or two keys could have the same keyid if someone chooses to assign "human readable" identifiers ("key-alice"). So far, we're just trusting the keyid value presented in the signature object to make the call.
2. Since keyids are optional, we can't even use the current "trust the keyid in the signature object" if we have two signatures from the same key, one with keyid and one without.

I think, then, trying to calculate thresholds and handling duplicate signatures gets very opinionated, and it's best to leave it to the consumers of the library, especially since each one may have their own take on threshold values. If we do want to, we'd have to map every key to every signature (which we more or less do when there are no keyids), calculate a consistent keyid value for every key, and then use that to track duplicates, all of which I think may be overkill. What do you think?

Reply:
I don't mind getting input from @lukpueh. I am working parallel on payload parser on my local.

Reply:
@adityasaky makes some good points. Let's not check for duplicates here, but leave it to the consumer/verifier.

Regarding...
> I don't think an attacker can utilize that as for that he/she will need the private key used for signing (and if he has that he can break the system on multiple levels), so maybe I shouldn't worry.

I might be misunderstanding your thought, @MVrachev, but the premise of signature threshold verification is that the attacker can indeed control some private keys, but as long that number of keys is below the threshold the attacker can't compromise the data. So verifying the threshold does serve a purpose, but it has to happen at verification time (also see [my comment below](https://github.com/in-toto/securesystemslib/pull/2#issuecomment-1173996576)).